lib/web_ui/lib/src/engine/canvas_pool.dart

```diff
@@ -0,0 +1,761 @@
+    // towards the threshold. Setting width and height to zero tricks Webkit
+    // into thinking that this canvas has a zero size so it doesn't count it
+    // towards the threshold.
+    if (browserEngine == BrowserEngine.webkit && _canvas != null) {
+      _canvas.width = _canvas.height = 0;
+    }
+    _clearPool();
+  }
+
+  void _clearPool() {
+    if (_pool != null) {
+      for (html.CanvasElement c in _pool) {
+        if (browserEngine == BrowserEngine.webkit) {
+          c.width = c.height = 0;
+        }
+        c.remove();
+      }
+    }
+    _pool = null;
+  }
+}
+
+// Optimizes applying paint parameters to html canvas.
+//
+// See https://www.w3.org/TR/2dcontext/ for defaults used in this class
+// to initialize current values.
+//
+class ContextStateHandle {
+  html.CanvasRenderingContext2D context;
+  ContextStateHandle(this.context);
+  ui.BlendMode _currentBlendMode = ui.BlendMode.srcOver;
+  ui.StrokeCap _currentStrokeCap = ui.StrokeCap.butt;
+  ui.StrokeJoin _currentStrokeJoin = ui.StrokeJoin.miter;
+  Object _currentFillStyle;
+  Object _currentStrokeStyle;
+  double _currentLineWidth = 1.0;
+  String _currentFilter = 'none';
+
+  set blendMode(ui.BlendMode blendMode) {
+    if (blendMode != _currentBlendMode) {
+      _currentBlendMode = blendMode;
+      context.globalCompositeOperation =
+          _stringForBlendMode(blendMode) ?? 'source-over';
+    }
+  }
+
+  set strokeCap(ui.StrokeCap strokeCap) {
+    strokeCap ??= ui.StrokeCap.butt;
+    if (strokeCap != _currentStrokeCap) {
+      _currentStrokeCap = strokeCap;
+      context.lineCap = _stringForStrokeCap(strokeCap);
+    }
+  }
+
+  set lineWidth(double lineWidth) {
+    if (lineWidth != _currentLineWidth) {
+      _currentLineWidth = lineWidth;
+      context.lineWidth = lineWidth;
+    }
+  }
+
+  set strokeJoin(ui.StrokeJoin strokeJoin) {
+    strokeJoin ??= ui.StrokeJoin.miter;
+    if (strokeJoin != _currentStrokeJoin) {
+      _currentStrokeJoin = strokeJoin;
+      context.lineJoin = _stringForStrokeJoin(strokeJoin);
+    }
+  }
+
+  set fillStyle(Object colorOrGradient) {
+    if (!identical(colorOrGradient, _currentFillStyle)) {
+      _currentFillStyle = colorOrGradient;
+      context.fillStyle = colorOrGradient;
+    }
+  }
+
+  set strokeStyle(Object colorOrGradient) {
+    if (!identical(colorOrGradient, _currentStrokeStyle)) {
+      _currentStrokeStyle = colorOrGradient;
+      context.strokeStyle = colorOrGradient;
+    }
+  }
+
+  set filter(String filter) {
+    if (_currentFilter != filter) {
+      _currentFilter = filter;
+      context.filter = filter;
+    }
+  }
+
+  void paint(ui.PaintingStyle style) {
+    if (style == ui.PaintingStyle.stroke) {
+      context.stroke();
+    } else {
+      context.fill();
+    }
+  }
+
+  void reset() {
+    context.fillStyle = '';
+    _currentFillStyle = context.fillStyle;
+    context.strokeStyle = '';
+    _currentStrokeStyle = context.strokeStyle;
```

Comment:
Can you please leave a comment explaining why you have to write an empty string then read it back from `context` rather than assign an empty string to everything?

Reply:
Dnoe.